sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java

```diff
@@ -131,47 +141,130 @@ public ShardingUpsertExecutor(ClusterService clusterService,
         try {
-            while (batchIterator.moveNext()) {
-                onRow(row);
-                if (idxWithinBatch == batchSize) {
-                    CompletableFuture<BitSet> executeBatchFuture = execute(false);
-                    idxWithinBatch = 0;
-                    return executeBatchFuture.thenCompose(r -> consumeIterator(batchIterator));
+            while (true) {
+                if (indexInBulk == bulkSize) {
+                    if (tryExecuteBulk(batchIterator, false) == false) {
+                        collectingEnabled.set(false);
+                        if (isScheduledCollectionRunning.compareAndSet(false, true)) {
+                            try {
+                                scheduleConsumeIterator(batchIterator);
+                            } catch (EsRejectedExecutionException e) {
+                                isScheduledCollectionRunning.set(false);
+                                if (collectingEnabled.compareAndSet(false, true)) {
+                                    // re-acquired the ownership of the iterator consumer, so let's try to
+                                    // execute / schedule this bulk again (otherwise someone else owns the
+                                    // iterator and carries on)
+                                    continue;
+                                }
+                            }
+                        }
+                        return;
+                    }
+                }
+
+                if (batchIterator.moveNext()) {
+                    onRow(row);
+                    pendingRequestsCount.incrementAndGet();
+                } else {
+                    break;
                 }
             }
 
             if (batchIterator.allLoaded()) {
-                batchIterator.close();
-                return execute(true);
+                lastBulkScheduledToExecute = true;
+                execute(true).whenComplete(bulkExecutionCompleteListener(batchIterator));
             } else {
-                return batchIterator.loadNextBatch().thenCompose(r -> consumeIterator(batchIterator));
+                batchIterator.loadNextBatch().whenComplete((r, t) -> {
+                    if (t == null) {
+                        unsafeConsumeIterator(batchIterator);
+                    } else {
+                        result.completeExceptionally(t);
+                    }
+                });
             }
         } catch (Throwable t) {
             batchIterator.close();
-            return CompletableFutures.failedFuture(t);
+            result.completeExceptionally(t);
+        }
+    }
+
+    private void scheduleConsumeIterator(BatchIterator batchIterator) throws EsRejectedExecutionException {
+        scheduler.schedule(() -> {
+                isScheduledCollectionRunning.set(false);
+                safeConsumeIterator(batchIterator);
+            },
+            consumeIteratorDelays.next().getMillis(), TimeUnit.MILLISECONDS);
```

Comment:
We would stop the operation in that case. Maybe a default value is a good way to guard ourselves against that.